modules/net/calico: enable BGP networking

s-urbaniak opened the pull request:
Also don't deploy flannel CNI if using calico network policy.

Fixes #INST-62, #INST-142, #INST-524

alexsomesan commented on platforms/aws/tectonic.tf:
```diff
@@ -163,16 +163,19 @@ module "flannel-vxlan" {
   calico_cni_image   = "${var.tectonic_container_images["calico_cni"]}"
   cluster_cidr       = "${var.tectonic_cluster_cidr}"
-  enabled            = "${var.tectonic_calico_network_policy}"
+  enabled            = "${var.tectonic_calico_network_policy || var.tectonic_calico_bgp}"
```
This is how you spot a C veteran ;)

s-urbaniak commented:
setting `do-not-merge` tentatively to first coordinate manifest changes.

s-urbaniak commented:
@squat PTAL

aaronlevy commented on config.tf:
```diff
@@ -450,6 +450,29 @@ This should only be enabled on clusters that are meant to be short-lived to begi
+  default = false
+
+  description = <<EOF
+(optional) [ALPHA] If set to true, calico BGP networking will be configured using IPinIP encapsulation.
```
Do we need to have this as an option out-of-the-gate or is it something that a default will cover most users?

s-urbaniak replied:
see comments in https://github.com/coreos/tectonic-installer/pull/1993#issuecomment-332752160

aaronlevy commented on modules/net/flannel-vxlan/resources/manifests/kube-flannel.yaml:
```diff
@@ -71,7 +71,7 @@ spec:
       - name: install-cni
         image: ${flannel_cni_image}
-        command: ["/install-cni.sh"]
+        command: [${flannel_cni_command}]
```
This is more of a nit, as control of these manifests/options will be moved out of the installer soon, but just want to raise the discussion point:

I feel like this would be better served by maintaining separate manifests that are conditionally applied as a whole. While this keeps the cni plugins and config from being applied, it leaves all kinds of unnecessary cruft around (containers, mounts, env vars).

For example, if you choose flannel as your provider, there is a distinct set of manifests that relate to flannel. Same with Canal, same with Calico. Otherwise these parameter substitutions become very hard to maintain long-term.

But again, soon we'll be moving such that the installer will simply output the provider selected into a config, and the business logic of transforming that into distinct manifests won't live in terraform -- so I wouldn't block on changing this.

s-urbaniak replied:
agreed, see https://github.com/coreos/tectonic-installer/pull/1993/files#r141541166

aaronlevy commented on platforms/aws/tectonic.tf:
```diff
@@ -160,19 +160,23 @@ module "tectonic" {
-  flannel_cni_image = "${var.tectonic_container_images["flannel_cni"]}"
-  cluster_cidr      = "${var.tectonic_cluster_cidr}"
+  cluster_cidr        = "${var.tectonic_cluster_cidr}"
+  enabled             = "${!var.tectonic_calico_bgp}"
```
nit: not sure how terraform outputs information about module variables, but if it's not clear - might suggest prefix for this.

aaronlevy commented:
cc @dghubble as you've been looking more closely at calico support in bootkube

dghubble commented:
High level observation is that this seems a bit overly complicated.

* Having a variable for everything may not be advisable
* Calico IPIP defaults to true, which is probably what you want as the default
* I don't see MTU here, its rather important this be set correctly on each platform and configurable on bare-metal.

I can point you to how my own projects run Calico (default) or flannel for AWS, bare-metal, GCE, and Digital Ocean if you have an interest.

dghubble commented on modules/net/calico/assets.tf:
```diff
@@ -0,0 +1,27 @@
+    cluster_cidr              = "${var.cluster_cidr}"
+    cni_version               = "${var.cni_version}"
+    felix_ipinip_enabled      = "${var.ipinip_enabled ? "true" : "false"}"
+    host_cni_bin              = "/var/lib/cni/bin"
```
Why is Tectonic using a different path from upstream?

s-urbaniak replied:
xref'ing the original discussion https://github.com/coreos/tectonic-installer/pull/780/files#r126188220

dghubble replied:
hm, ok. Thanks for the link. Looks like a conscious choice you guys are making for organization.

dghubble commented on modules/net/calico/assets.tf:
```diff
@@ -0,0 +1,27 @@
+    felix_ipinip_enabled      = "${var.ipinip_enabled ? "true" : "false"}"
+    host_cni_bin              = "/var/lib/cni/bin"
+    kube_apiserver_url        = "${var.kube_apiserver_url}"
+    log_level                 = "${var.log_level}"
```
Is this hoping every app supports the standard log levels?

diegs replied:
I don't think this should be a parameter.

s-urbaniak replied:
Unrelated to this PR due to the mechanical file move, but I am more than happy to pin this to the default value `WARNING` in this PR.

dghubble commented on modules/net/calico/assets.tf:
```diff
@@ -0,0 +1,27 @@
+    calico_cni_image          = "${var.calico_cni_image}"
+    calico_image              = "${var.calico_image}"
+    cluster_cidr              = "${var.cluster_cidr}"
+    cni_version               = "${var.cni_version}"
```
Why is this even configurable? It probably better to set what you'll actually be supporting.

s-urbaniak replied:
This is unrelated to this PR, as this is a mechanical file rename, but I am more than happy to pin this to `0.3.0` in this PR.

dghubble replied:
Sounds good. You'll need to use CNI 0.3.1 when bumping to flannel-cni:v0.3.0, unless you override the default CNI config.

https://github.com/kubernetes-incubator/bootkube/pull/697#issuecomment-329617739
https://github.com/kubernetes-incubator/bootkube/pull/697#issuecomment-329839399
https://github.com/coreos/flannel-cni/releases/tag/v0.3.0

s-urbaniak replied:
I'd leave the bump to flannel-cni from v0.2.0 to v0.3.0 as a separate PR and leave this as a mechanical refactor if you don't mind (one change at a time). I created INST-517 to track the flannel 0.3.0 bump.

dghubble commented on modules/net/calico/assets.tf:
```diff
@@ -0,0 +1,27 @@
+
+    calico_cni_image          = "${var.calico_cni_image}"
+    calico_image              = "${var.calico_image}"
+    cluster_cidr              = "${var.cluster_cidr}"
```
In other places we've been clarifying `pod_cidr` or `service_cidr`

s-urbaniak replied:
`tectonic_cluster_cidr` is a top-level variable injected into this value. `pod_cidr` is not being used in the installer. While I agree that the term `pod_cidr` is the better term, this question is unrelated to this PR.

dghubble replied:
ok

dghubble commented on modules/net/flannel-vxlan/assets.tf:
```diff
@@ -2,10 +2,11 @@ data "template_file" "flannel" {
-    cluster_cidr      = "${var.cluster_cidr}"
-    host_cni_bin      = "/var/lib/cni/bin"
+    cluster_cidr        = "${var.cluster_cidr}"
+    flannel_cni_command = "${var.install_cni_enabled ? "\"/install-cni.sh\"" : "\"/bin/sh\", \"-c\", \"/bin/echo CNI installation skipped; while :; do sleep 3600; done;\""}"
```
Feels wrong. At what point is everything just a variable and the installer just a general template? Looks like this is done to save one manifest, is it worth it?

s-urbaniak replied:
Relates to https://github.com/coreos/tectonic-installer/pull/1993/files#r141493957: yes, this was to reduce the amount of manifests, as optional modules in terraform are effectively non-existent. We do have some (hacky) HCL/ICL tricks in place to cope with this, i.e. calculating hashes of module IDs to enforce happens-before invariants, but it is not pretty.

Nevertheless I am happy to introduce a dedicated `modules/net/calico-bgp` manifest and trade TF complexity, as I agree with @aaronlevy that this leaves unnecessary cruft around and in the long-term this manifest will disappear from this repository (and thus also quite some TF hacks enforcing invariants).

s-urbaniak commented:
@dghubble @aaronlevy 

> Calico IPIP defaults to true, which is probably what you want as the default

(related to https://github.com/coreos/tectonic-installer/pull/1993#discussion_r141492701)

According to https://docs.projectcalico.org/v2.4/usage/configuration/ip-in-ip IP-in-IP is an escape hatch in case (quote) "your network fabric performs source/destination address checks and drops traffic when those addresses are not recognized, it may be necessary to enable IP-in-IP encapsulation of the inter-workload traffic."

I don't have strong feelings to leave this as a variable or defaulting this to `true`, from the documentation ip-in-ip seemed to me to be a workaround in restricted networks. I'll default it to `true`, as going along with the default seems to be the consensus.

> I don't see MTU here, its rather important this be set correctly on each platform and configurable on bare-metal. Be sure to run performance benchmarks.

Very good catch, I'll add a configuration option for MTU, thanks!

Running performance benchmarks per platform is indeed a very good idea. We don't have such a suite in place now though for the time being, but I'll make sure to create an issue. For the time being I agree to make MTU configurable.

dghubble commented:
> I'll add a configuration option for MTU, thanks

Thanks! In my own testing across various platforms, I've found the pod to pod network performance on AWS tends to be rather poor unless the jumbo frame MTU is chosen.

rithujohn191 commented on Documentation/variables/config.md:
```diff
@@ -13,6 +13,8 @@ This document gives an overview of variables used in all platforms of the Tecton
 | tectonic_ca_cert | (optional) The content of the PEM-encoded CA certificate, used to generate Tectonic Console's server certificate. If left blank, a CA certificate will be automatically generated. | string | `` |
 | tectonic_ca_key | (optional) The content of the PEM-encoded CA key, used to generate Tectonic Console's server certificate. This field is mandatory if `tectonic_ca_cert` is set. | string | `` |
 | tectonic_ca_key_alg | (optional) The algorithm used to generate tectonic_ca_key. The default value is currently recommended. This field is mandatory if `tectonic_ca_cert` is set. | string | `RSA` |
+| tectonic_calico_bgp | (optional) [ALPHA] If set to true, calico BGP networking support will be deployed. WARNING: Enabling an alpha feature means that future updates may become unsupported. This should only be enabled on clusters that are meant to be short-lived to begin validating the alpha feature. Note that this feature is currently only known to work on baremetal installations. | string | ` ...
```
Would it be reasonable to rename all the `calico_bgp` variables to just `calico`, since calico can only use `BGP` seems like an overkill

dghubble commented on platforms/openstack/neutron/main.tf:
```diff
@@ -286,17 +286,21 @@ module "dns" {
   cluster_cidr       = "${var.tectonic_cluster_cidr}"
-  enabled            = "${var.tectonic_calico_network_policy}"
+  enabled            = "${var.tectonic_calico_network_policy || var.tectonic_calico_bgp}"
+  bgp_enabled        = "${var.tectonic_calico_bgp}"
```
By `bgp_enabled`, are you intending to mean Calic's special `policy-only` mode that is run atop a cluster with an existing flannel setup? We might want to make this more clear because at first I'm wondering, "BGP, of course BGP is enabled, that's how Calico (proper) exchanges routes"

dghubble commented:
I just noticed Tectonic is using Calico v2.4.1. I think you'll run into TPR issues on Kubernetes v1.8 unless I've missed a change somewhere.

s-urbaniak commented:
ok to test

s-urbaniak commented:
@dghubble k8s 1.8 is out of scope for this PR.

s-urbaniak commented:
@dghubble @rithujohn191 @aaronlevy thanks a lot for the valuable input! I refactored the requested changes. There are now three separate manifest sets in `modules/net`:

- `flannel-vxlan`: contains the flannel-only manifests (the current default)
- `canal`: contains the calico network policy manifest and the flannel-without-cni manifest (effectively fixing INST-142)
- `calico`: contains the calico BGP manifest with configurability of MTUs for baremetal only. Once @cpanato is ready with the bare-metal test refactoring, we should add a dedicated test for this mode. This is tracked in INST-518.

The network flavor can be selected using the top-level `tectonic_networking` variable which can have the string values `flannel`, `canal`, or `calico`.

alexsomesan commented:
Tested on Azure with Canal - works as expected.

aaronlevy commented on config.tf:
```diff
@@ -452,13 +452,20 @@ Specifies the RFC2136 Dynamic DNS server key secret.
   description = <<EOF
-(optional) [ALPHA] If set to true, calico network policy support will be deployed.
-WARNING: Enabling an alpha feature means that future updates may become unsupported.
+(optional) [ALPHA] Configures the network to be used in Tectonic. One of the following values can be used:
```
flannel isn't alpha, just canal/calico. One option might be to use "alpha" as part of the identifier:

```
flannel
alpha-canal
alpha-calico
```

s-urbaniak replied:
My suggestion would be to leave these values. Once we introduce a tectonic-network-operator, we can pass this setting 1:1 in a ConfigMap without changing values.

Instead I'll add the `[ALPHA]` label to the bullet points instead of the top-level description.

aaronlevy commented on config.tf:
```diff
@@ -452,13 +452,20 @@ Specifies the RFC2136 Dynamic DNS server key secret.
+- "calico": enables Calico (BGP) based networking.
+
 This should only be enabled on clusters that are meant to be short-lived to begin validating the alpha feature.
+Note that this feature is currently only known to work on baremetal installations.
```
The way this is structured with flannel above is confusing. Should be explicit (canal/calico are only known to work on baremetal installations). Also, why is this the case?

dghubble replied:
I think he just means that's what they tested. Calico can work quite fine on cloud platforms.

dghubble replied:
I also think some of the language here is confusing. "native" flannel? What do you mean by "native". Flannel is an xvlan overlay, Calico is IP or IPIP routing with BGP for determining routes. If any of them are described as "native" its Calico.

For network policy, maybe include that in parenthesis (implements network polocy) since Calico and Canal both have it, but that's not so clear to the reader.

aaronlevy commented on modules/net/calico/assets.tf:
```diff
@@ -0,0 +1,19 @@
+    cluster_cidr       = "${var.cluster_cidr}"
+    kube_apiserver_url = "${var.kube_apiserver_url}"
+    mtu                = "${var.mtu}"
+    mtu_ipinip         = "${var.mtu_ipinip}"
```
is there ever a reason to have different MTUs for each of these modes?

aaronlevy replied:
Also, maybe instead we just set a sane default and if the user wants to modify it they can go do that directly in the manifest. cc @dghubble 

It's a slippery slope to just plumb through every option (we should have strong reasons for doing this vs, "if you need a different MTU - go change it in the manifest).

dghubble replied:
The correct value varies by platform. On AWS, the correct value varies by instance type. You could use the lowest value as an out-of-box default, but it sacrifices performance.

I'd recommend fixing it on platforms where there is a single correct value and allowing it to be customized only where users need to think about it (AWS). [Performance stats](
https://typhoon.psdn.io/topics/performance/#network-performance).

aaronlevy replied:
@dghubble is there any reason to have both `mtu` and `mtu_ipinip` two configuration items or can we just set it once?

dghubble replied:
When I commended, I thought there was just one MTU setting, but maybe I missed this.

No, I don't think you want to allow users to customize both. Its recommended they match and you can reduce complexity and prevent misconfigurations by not allowing them to differ unless a good reason comes along: "Typically the MTU for your workload interfaces [i.e. pods] should match the network MTU. If you need IP-in-IP then the MTU size for both the workload and tunnel interfaces should be 20 bytes less than the network MTU for your network."

In general, I think there are really only two new variables users should touch for networking: `network` (like flannel, calico, etc.) and `network_mtu`.

s-urbaniak replied:
ack, I'll set the `mtu_ipinip=mtu-20 bytes` internally then.

dghubble replied:
I think your variables here correspond to the workload and tunnel MTUs and should match, unless I've misunderstood something. The network MTU is whatever the platform actually offers.

aaronlevy commented on modules/net/calico/assets.tf:
```diff
@@ -0,0 +1,19 @@
+    calico_cni_image   = "${var.container_images["calico_cni"]}"
+    calico_image       = "${var.container_images["calico"]}"
+    cluster_cidr       = "${var.cluster_cidr}"
+    kube_apiserver_url = "${var.kube_apiserver_url}"
```
This is rather unfortunate. Are we sure it's needed? Ideally we would just use the service (`https://kubernetes` or `https://kubernetes.default.svc` etc). Or because we're deploying calico as a pod -- apiserver addressability should be known / injected via the service account. Probably worth at least looking into if this is actually needed.

dghubble replied:
Why can't you use the usual `https://__KUBERNETES_SERVICE_HOST__:__KUBERNETES_SERVICE_PORT__`

s-urbaniak replied:
This has been taken over 1:1 from the existing setup on master: https://github.com/coreos/tectonic-installer/blob/35c5377/modules/net/calico-network-policy/resources/manifests/kube-calico.yaml#L33.

I do agree that accessing the API server via the service is more optimal. Having said that I was under the impression, that (mainly due to the service IP flapping which has been addressed in https://github.com/kubernetes/kubernetes/pull/51698 coming in k8s >= 1.9) we prefer communicating via the external API endpoint rather than the internal service IP.

Being an alpha feature I suggest to address this as a follow-up. I filed INST-524 to track this.

aaronlevy replied:
I would prefer that we just fix this now. We don't want to be plumbing this through then remembering to remove it later - we know it's not the approach we should be using.

aaronlevy replied:
Also it looks like the calico docs use @dghubble's suggestion - so we should probably just switch to that.

s-urbaniak replied:
ack, no problem, I'll switch to the service IP then.

mxinden commented:
Ok to test

s-urbaniak commented:
ok to test

s-urbaniak commented:
@aaronlevy @dghubble PTAL at the corrected doc. If you have strong feelings, I can fix the comment in https://github.com/coreos/tectonic-installer/pull/1993#discussion_r143285288 in this PR or (as explained above) as a follow-up. If you have no objections, I'd go ahead and merge once green and move on with the bump in the k8s 1.8 context.

s-urbaniak commented:
@dghubble addressed feedback, PTAL. Final question from my side: I seem to have issues (locally) when testing with IPinIP enabled. When leaving disabled, all works fine but with IPinIP enabled I don't seem to have pod-to-pod connectivity. Do you have tips on debugging this?

Having said that (also after discussing with @squat) we still seem to be of the opinion to have IPinIP configurable. As pointed out above this also seems to be the outcome of Calico's docs, stating to enable IPinIP if (and only if) the network topology is restricted (i.e. AWS). I still do believe that customers might want to opt-out/opt-in to IPinIP encapsulation in a per-use-case basis.

squeed commented on modules/net/calico/resources/manifests/kube-calico.yaml:
```diff
@@ -0,0 +1,279 @@
+            # Enable IPIP
+            - name: CALICO_IPV4POOL_IPIP
+              value: "always"
+            - name: FELIX_IPINIPENABLED
```
Careful, enabling IPinIP is not a safe default:
* It doesn't work in Azure
* It doesn't make sense for bare-metal, from a performance and reachability perspective.

dghubble replied:
It is the upstream default and supposed to be the lowest common denominator that works despite some cloud providers dropping unrecognized packets. I agree its not required on bare-metal, I've left it on in my own clusters, the performance difference is quite minimal.

I can't speak to Azure at all.

sym3tri replied:
I propose we move forward with this PR as-is, and keep the setting as `true`. 
Rationale:

- This is an alpha feature. We can iterate and make additional changes before it goes to beta.
- It is only intended for _evaluation_ and _testing_ for baremetal platforms.
- The upstream default is `true`.
- Turning it off on baremetal seems like a performance optimization.
- Most importantly, this is not the final design. We have an open issues that must be addressed before this feature moves to beta. We can have ongoing discussions there on the design going forward (INST-529, INST-435) and weather or not to make this configurable etc.

rithujohn191 commented on modules/net/calico/outputs.tf:
```diff
@@ -0,0 +1,7 @@
+output "id" {
+  value = "${var.enabled ? "${sha1("${join(" ", local_file.calico.*.id)}")}" : "# calico disabled"}"
```
a little confused about the need for this

s-urbaniak replied:
indeed it is, the tldr is that TF cannot compute dependencies between data sources and modules as a whole execution unit:
https://github.com/coreos/tectonic-installer/blob/d301ea000e5891493440faf674f318f1e700c5cf/platforms/aws/tectonic.tf#L179-L189

s-urbaniak commented:
I created INST-529 to clarify IPinIP configurability. Since this is alpha we can proceed and address this question as a follow-up.

PTAL @alexsomesan for final approval.
